From de5df7ba9255e866f1cf3a5bb1817772ea54f56d Mon Sep 17 00:00:00 2001
From: Haoran Wang <b50027@freescale.com>
Date: Wed, 15 Jul 2015 16:53:26 +0800
Subject: [PATCH] MA-6854-1 Add PM_IGNORE_NOTIFY flag to mmc pm flag

BCM4339 device need PM_IGNORE_NOTIFY flag in mmc driver.
Added the property in dts and add the flag when boot up.

This patch will fix the BCM4339 suspend panic issue like:
    [<c05f567c>] (sdio_claim_host) from [<bf08bdb4>]
(sdioh_request_byte+0x1ac/0x28c [bcmdhd])
    [<bf08bdb4>] (sdioh_request_byte [bcmdhd]) from [<bf08beb8>]
(sdioh_cfg_write+0x24/0x2c [bcmdhd])
    [<bf08beb8>] (sdioh_cfg_write [bcmdhd]) from [<bf08a980>]
(bcmsdh_cfg_write+0x44/0xc0 [bcmdhd])
    [<bf08a980>] (bcmsdh_cfg_write [bcmdhd]) from [<bf094db4>]
(dhdsdio_clk_kso_enab+0x34/0x160 [bcmdhd])
    [<bf094db4>] (dhdsdio_clk_kso_enab [bcmdhd]) from [<bf094fcc>]
(dhdsdio_clk_devsleep_iovar+0xec/0x578 [bcmdhd])
    [<bf094fcc>] (dhdsdio_clk_devsleep_iovar [bcmdhd]) from [<bf0974d0>]
(dhdsdio_bussleep+0x2e0/0x4bc [bcmdhd])
    [<bf0974d0>] (dhdsdio_bussleep [bcmdhd]) from [<bf09ddec>]
(dhd_bus_txctl+0x414/0x560 [bcmdhd])
    [<bf09ddec>] (dhd_bus_txctl [bcmdhd]) from [<bf08d068>]
(dhdcdc_msg+0x48/0x7c [bcmdhd])
    [<bf08d068>] (dhdcdc_msg [bcmdhd]) from [<bf08d420>]
(dhd_prot_ioctl+0x23c/0x628 [bcmdhd])
    [<bf08d420>] (dhd_prot_ioctl [bcmdhd]) from [<bf04c868>]
(dhd_wl_ioctl+0x4c/0xc8 [bcmdhd])
    [<bf04c868>] (dhd_wl_ioctl [bcmdhd]) from [<bf057938>]
(dhd_ioctl_entry_local+0x58/0xe4 [bcmdhd])
    [<bf057938>] (dhd_ioctl_entry_local [bcmdhd]) from [<bf08853c>]
(wldev_ioctl+0x40/0x48 [bcmdhd])
    [<bf08853c>] (wldev_ioctl [bcmdhd]) from [<bf08630c>]
(wl_cfgp2p_bss_isup+0x40/0x108 [bcmdhd])
    [<bf08630c>] (wl_cfgp2p_bss_isup [bcmdhd]) from [<bf07ec54>]
(wl_cfg80211_down+0x7ac/0x88c [bcmdhd])
    [<bf07ec54>] (wl_cfg80211_down [bcmdhd]) from [<bf0528a4>]
(dhd_detach+0x48/0x4ec [bcmdhd])

Signed-off-by: Haoran Wang <b50027@freescale.com>
---
 arch/arm/boot/dts/imx7d-sdb.dts    | 1 +
 drivers/mmc/host/sdhci-esdhc-imx.c | 5 +++++
 2 files changed, 6 insertions(+)

diff --git a/arch/arm/boot/dts/imx7d-sdb.dts b/arch/arm/boot/dts/imx7d-sdb.dts
index 5b6439a..ca8eb46 100644
--- a/arch/arm/boot/dts/imx7d-sdb.dts
+++ b/arch/arm/boot/dts/imx7d-sdb.dts
@@ -1122,6 +1122,7 @@
 	keep-power-in-suspend;
 	tuning-step = <2>;
 	wifi-host;
+    pm-ignore-notify;
 	status = "okay";
 };
 
diff --git a/drivers/mmc/host/sdhci-esdhc-imx.c b/drivers/mmc/host/sdhci-esdhc-imx.c
index ea03809..5a8804e 100644
--- a/drivers/mmc/host/sdhci-esdhc-imx.c
+++ b/drivers/mmc/host/sdhci-esdhc-imx.c
@@ -1057,6 +1057,11 @@ sdhci_esdhc_imx_probe_dt(struct platform_device *pdev,
 		dev_info(mmc_dev(host->mmc), "assigned as wifi host\n");
 	}
 
+	if (of_get_property(np, "pm-ignore-notify", NULL)) {
+		host->mmc->pm_caps |= MMC_PM_IGNORE_PM_NOTIFY;
+		host->mmc->pm_flags = host->mmc->pm_caps;
+	}
+
 	return 0;
 }
 #else
-- 
1.8.0

